File: client/src/api/qa.js
/** QA pipeline API - contract: server/app/schemas/qa.py (BE: Person 3, FE: Person 2). */
import { isDemoMode } from '@/config';
import qaStepsMock from '@/mocks/qa_steps_mock.json';
import qaRunMock from '@/mocks/qa_run_mock.json';
import { http } from './http';
import { demo } from './demo';

let demoSteps = null;

export const qaApi = {
  getSteps(projectId) {
    if (isDemoMode) {
      demoSteps ??= qaStepsMock.steps;
      return demo({ project_id: projectId, steps: demoSteps });
    }
    return http.get(`/api/v1/projects/${projectId}/qa/steps`);
  },

  saveSteps(projectId, steps) {
    if (isDemoMode) {
      demoSteps = steps.map((s, i) => ({ ...s, order: i + 1 }));
      return demo({ project_id: projectId, steps: demoSteps });
    }
    return http.post(`/api/v1/projects/${projectId}/qa/steps`, { steps });
  },

  run(projectId, stepIds = null) {
    if (isDemoMode) return demo({ ...qaRunMock, project_id: projectId }, 2500);
    return http.post(`/api/v1/projects/${projectId}/qa/run`, { steps: stepIds });
  },
};

File: server/app/api/qa.py
"""QA pipeline endpoints. Owner: Person 3 | BL-QA-01, BL-QA-02."""

from fastapi import APIRouter

from app.schemas.qa import QARunRequest, QARunResponse, QAStepsResponse, QAStepsUpdateRequest
from app.services import qa_service

router = APIRouter(prefix="/api/v1/projects/{project_id}/qa", tags=["qa"])


@router.get("/steps", response_model=QAStepsResponse)
async def get_steps(project_id: str) -> QAStepsResponse:
    return QAStepsResponse(project_id=project_id, steps=qa_service.get_steps(project_id))


@router.post("/steps", response_model=QAStepsResponse)
async def save_steps(project_id: str, req: QAStepsUpdateRequest) -> QAStepsResponse:
    return QAStepsResponse(
        project_id=project_id, steps=qa_service.save_steps(project_id, req.steps)
    )


@router.post("/run", response_model=QARunResponse)
async def run_qa(project_id: str, req: QARunRequest | None = None) -> QARunResponse:
    return await qa_service.run(project_id, req.steps if req else None)
